<launch>
  <arg name="mav_name" default="hummingbird"/>
  <arg name="world_name" default="arena"/>
  <arg name="enable_logging" default="false" />
  <arg name="enable_ground_truth" default="true" />
  <arg name="log_file" default="$(arg mav_name)" />
  <arg name="debug" default="false"/>
  <arg name="gui" default="true"/>
  <arg name="paused" default="false"/>
  <!-- The following line causes gzmsg and gzerr messages to be printed to the console
      (even when Gazebo is started through roslaunch) -->
  <arg name="verbose" default="false"/>
  <arg name="rvizconfig" default="$(find quad_control)/drone.rviz" />
  <arg name="use_sim_time" default="true" />
  
  <!-- Controller parameters -->
  <arg name="rate"  default="1000" />
  <arg name="mass"  default="0.68" />
  <arg name="Ixx"  default="0.007" />
  <arg name="Iyy"  default="0.007" />
  <arg name="Izz"  default="0.012" />

  <env name="GAZEBO_MODEL_PATH" value="${GAZEBO_MODEL_PATH}:$(find rotors_gazebo)/models"/>
  <env name="GAZEBO_RESOURCE_PATH" value="${GAZEBO_RESOURCE_PATH}:$(find rotors_gazebo)/models"/>
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(find quad_control)/worlds/$(arg world_name).world" />
    <arg name="debug" value="$(arg debug)" />
    <arg name="paused" value="$(arg paused)" />
    <arg name="gui" value="$(arg gui)" />
    <arg name="verbose" value="$(arg verbose)"/>
  </include>

  <group ns="$(arg mav_name)">
    <include file="$(find rotors_gazebo)/launch/spawn_mav.launch">
      <arg name="mav_name" value="$(arg mav_name)" />
      <arg name="model" value="$(find rotors_description)/urdf/mav_generic_odometry_sensor.gazebo" />
      <arg name="enable_logging" value="$(arg enable_logging)" />
      <arg name="enable_ground_truth" value="$(arg enable_ground_truth)" />
      <arg name="log_file" value="$(arg log_file)"/>
    </include>
  
    <include file="$(find quad_control)/launch/plot_juggler.launch"/>

    <node name="odometry_node" type="odometry_node" pkg="quad_control">
      <param  name="rate"           type="double" value="$(arg rate)"/>
      <remap  from="/odometry"      to="/$(arg mav_name)/ground_truth/odometryNED"/>
      <remap  from="/p"             to="/$(arg mav_name)/odometryNED/p"/>
      <remap  from="/eta"           to="/$(arg mav_name)/odometryNED/eta"/>
    </node> 

    <node name="trajectory_node" type="trajectory_node" pkg="quad_control">
      <param  name="threshold"      type="double" value="0.2"/>
      <param  name="ka"             type="double" value="0.5"/>
      <param  name="kf_d"           type="double" value="10"/>
      <param  name="rate"           type="double" value="$(arg rate)"/>
      <remap  from="/p"             to="/$(arg mav_name)/odometryNED/p"/>
      <remap  from="/traj/p"        to="/$(arg mav_name)/trajectory/p"/>
      <remap  from="/traj/yaw"      to="/$(arg mav_name)/trajectory/yaw"/>
    </node>

    <node name="outer_loop_node" type="outer_loop_node" pkg="quad_control">
      <param  name="mass"           type="double" value="$(arg mass)"/>
      <param  name="Kp_x"           type="double" value="100"/>
      <param  name="Kp_y"           type="double" value="100"/>
      <param  name="Kp_z"           type="double" value="100"/>
      <param  name="Kd_x"           type="double" value="10"/>
      <param  name="Kd_y"           type="double" value="10"/>
      <param  name="Kd_z"           type="double" value="10"/>
      <remap  from="/traj_topic"    to="/$(arg mav_name)/trajectory/p"/>
      <remap  from="/odom_topic"    to="/$(arg mav_name)/odometryNED/p"/>
      <remap  from="/est_topic"     to="/$(arg mav_name)/estimator"/>
      <remap  from="/mu_hat"        to="/$(arg mav_name)/mu_hat"/>
    </node>

    <node name="eta_reference_node" type="eta_reference_node" pkg="quad_control">
      <param  name="mass"           type="double"  value="$(arg mass)"/>
      <param  name="rate"           type="double" value="$(arg rate)"/>
      <param  name="kf_d"           type="double"  value="10"/>
      <param  name="kf_dd"          type="double"  value="10"/>
      <remap  from="/sub_topic"     to="/$(arg mav_name)/mu_hat"/>
      <remap  from="/traj_topic"    to="/$(arg mav_name)/trajectory/yaw"/>
      <remap  from="/eta_ref_topic" to="/$(arg mav_name)/eta_reference"/>
    </node>

    <node name="inner_loop_node" type="inner_loop_node" pkg="quad_control">
      <param  name="mass"           type="double"  value="$(arg mass)"/>
      <param  name="Ixx"            type="double"  value="$(arg Ixx)"/>
      <param  name="Iyy"            type="double"  value="$(arg Iyy)"/>
      <param  name="Izz"            type="double"  value="$(arg Izz)"/>
      <param  name="v"              type="double"  value="20"/>
      <param  name="Kp_x"           type="double" value="0.6"/>
      <param  name="Kp_y"           type="double" value="0.6"/>
      <param  name="Kp_z"           type="double" value="1"/>
      <param  name="Kd_x"           type="double" value="0.3"/>
      <param  name="Kd_y"           type="double" value="0.3"/>
      <param  name="Kd_z"           type="double" value="0.5"/>
      <remap  from="/eta_ref_topic" to="/$(arg mav_name)/eta_reference"/>
      <remap  from="/odom_topic"    to="/$(arg mav_name)/odometryNED/eta"/>
      <remap  from="/mu_topic"      to="/$(arg mav_name)/mu_hat" />
      <remap  from="/est_topic"     to="/$(arg mav_name)/estimator" />
      <remap  from="/pub_topic"     to="/$(arg mav_name)/command/wrenchNED"/>
    </node>
<!--
    <node name="estimator_node" type="estimator_node" pkg="quad_control">
      <param name="mass"            type="double"  value="$(arg mass)"/>
      <param name="k0"              type="double"  value="10"/>
      <param name="rate"            type="double"  value="$(arg rate)"/>
      <param  name="Ixx"            type="double"  value="$(arg Ixx)"/>
      <param  name="Iyy"            type="double"  value="$(arg Iyy)"/>
      <param  name="Izz"            type="double"  value="$(arg Izz)"/>
      <remap  from="/eta"           to="/$(arg mav_name)/odometryNED/eta" />
      <remap  from="/p"             to="/$(arg mav_name)/odometryNED/p"/>
      <remap  from="/cmd"           to="/$(arg mav_name)/command/wrenchNED"/>
      <remap  from="/est_topic"     to="/$(arg mav_name)/estimator"/>
    </node>
-->
<!--  
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" />
    <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" />

    <node pkg="tf" type="static_transform_publisher" name="maptf" args="0 0 0 0 0 0  world /map  100"/>
    <node pkg="tf" type="static_transform_publisher" name="wNED" args="0 0 0 1.57079 0 3.14159  world worldNED  100"/>
    <node pkg="tf" type="static_transform_publisher" name="bodyNED" args="0 0 0 1.57079 0 3.14159  hummingbird/base_link hummingbird/base_linkNED  100"/>
    
     <node name="rviz" pkg="rviz" type="rviz" args="-d $(arg rvizconfig)"  />
    -->
  </group>
</launch>
